In [1]:
import pandas as pd
import numpy as np

In [3]:
#loading the raw data
file_path = "../data/Watershed_Water_Quality_-_Hydrology.csv"
df = pd.read_csv(file_path)

/var/folders/2t/5z71t9kj2yb0syg9m7v2fbnr0000gn/T/ipykernel_815/1964506953.py:3: DtypeWarning: Columns (3,5) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(file_path)


In [40]:
#initial inspection
print("Initial Data Preview: ")
df.head()

Initial Data Preview: 


Sample Id Sample Site Sample Date Sample Time          Analyte Status  \
0  C-185028          S4  1987-01-07         NaT      Temperature    NaN   
1  C-186342         S6I  1987-01-07         NaT      Temperature    NaN   
2  C-187744          S8  1987-01-07         NaT      Temperature    NaN   
3  C-185028          S4  1987-01-07         NaT  Scent Character    NaN   
4  C-186342         S6I  1987-01-07         NaT  Scent Character    NaN   

   Final Result Units Stream Group  
0      1.111110     C    Schoharie  
1      0.555556     C    Schoharie  
2      2.222220     C    Schoharie  
3           NaN   NaN    Schoharie  
4           NaN   NaN    Schoharie

In [41]:
df.tail()

Sample Id Sample Site Sample Date         Sample Time  \
1664702  E-2301489-01       BOYDR  2023-04-25 1900-01-01 09:52:00   
1664703  E-2301489-03   CROSSRVVC  2023-04-25 1900-01-01 08:40:00   
1664704  E-2301489-03   CROSSRVVC  2023-04-25 1900-01-01 08:40:00   
1664705  E-2301489-01       BOYDR  2023-04-25 1900-01-01 09:52:00   
1664706  E-2301489-02  CROFALLSVC  2023-04-25 1900-01-01 09:16:00   

                      Analyte Status  Final Result     Units  \
1664702  Specific Conductance   Done        237.00  umhos/cm   
1664703  Specific Conductance   Done        263.00  umhos/cm   
1664704             Turbidity   Done          1.60       NTU   
1664705             Turbidity   Done          0.96       NTU   
1664706             Turbidity   Done          0.99       NTU   

              Stream Group  
1664702        West Branch  
1664703  Delaware Aqueduct  
1664704  Delaware Aqueduct  
1664705        West Branch  
1664706  Delaware Aqueduct

In [42]:
df.columns

Index(['Sample Id', 'Sample Site', 'Sample Date', 'Sample Time', 'Analyte',
       'Status', 'Final Result', 'Units', 'Stream Group'],
      dtype='object')

In [43]:
# I don't want the unwanted column in the begining
df = df.reset_index(drop=True)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1664707 entries, 0 to 1664706
Data columns (total 9 columns):
 #   Column        Non-Null Count    Dtype 
---  ------        --------------    ----- 
 0   Sample Id     1664707 non-null  object
 1   Sample Site   1664707 non-null  object
 2   Sample Date   1664707 non-null  object
 3   Sample Time   1244822 non-null  object
 4   Analyte       1664707 non-null  object
 5   Status        298185 non-null   object
 6   Final Result  1664676 non-null  object
 7   Units         1528186 non-null  object
 8   Stream Group  1661495 non-null  object
dtypes: object(9)
memory usage: 114.3+ MB


In [20]:
#as every column have Dtype as object, it would be great to convert the types accordingly

#convert date and time
df['Sample Date'] = pd.to_datetime(df['Sample Date'], errors = 'coerce')
df['Sample Time'] = pd.to_datetime(df['Sample Time'], errors = 'coerce', format = '%H:%M:%S.%f')

#converting numerical columns:
df['Final Result'] = pd.to_numeric(df['Final Result'], errors = 'coerce')

#converting categorical columns
catergorical_cols = ['Sample Id', 'Sample Site', 'Analyte', 'Status', 'Units', 'Stream Group']
df[categorical_cols] = df[categorical_cols].astype('category')

#display the updated data types
df.dtypes

Sample Id               object
Sample Site           category
Sample Date     datetime64[ns]
Sample Time     datetime64[ns]
Analyte               category
Status                category
Final Result           float64
Units                 category
Stream Group          category
dtype: object

In [21]:
print(f"Initial dataset shape: {df.shape}")

Initial dataset shape: (1664707, 9)


In [22]:
df.describe()

Sample Date                    Sample Time  \
count                        1664707                         284857   
mean   2001-09-16 02:26:37.327577856  1900-01-01 11:01:51.187156480   
min              1987-01-05 00:00:00            1900-01-01 00:00:00   
25%              1994-09-12 00:00:00            1900-01-01 10:05:00   
50%              1999-11-15 00:00:00            1900-01-01 10:55:00   
75%              2007-10-01 00:00:00            1900-01-01 11:50:00   
max              2023-04-25 00:00:00            1900-01-01 23:56:00   
std                              NaN                            NaN   

       Final Result  
count  1.420296e+06  
mean   1.047976e+02  
min   -1.638890e+01  
25%    1.250000e+00  
50%    6.970000e+00  
75%    1.800000e+01  
max    1.470000e+06  
std    2.752474e+03

📊 Interpretation of df.describe() Results

1️⃣ Sample Date (Datetime)
Total Samples with Dates: 1,664,707 (all non-null).
Mean (Average Date): Sep 16, 2001 → Most samples are around this period.
Oldest Sample: Jan 5, 1987 → Data collection started in 1987.
Newest Sample: Apr 25, 2023 → Latest recorded sample.
    
Data Spread:
25% of data is from before Sep 12, 1994.
50% of data (median) is from before Nov 15, 1999.
75% of data is from before Oct 1, 2007.
    
Key Insight:
✅ Dataset covers a 36-year period (1987-2023), with more recent samples.
    
2️⃣ Sample Time (Datetime)
Total Samples with Time Recorded: 284,857 (~17% of total data).
Mean Time: 11:01 AM → Most samples were collected around this time.
Earliest Time: 00:00:00 (Midnight).
Latest Time: 23:56:00 (Near Midnight).
    
Data Spread:
25% of samples were taken before 10:05 AM.
50% (Median) were taken before 10:55 AM.
75% of samples were taken before 11:50 AM.
    
Key Insight:
✅ Most samples are collected in the morning, peaking before noon.
✅ Many missing timestamps (~83% of data).
    
3️⃣ Final Result (Float - Measurement Values)
Total Non-Null Values: 1,420,296 (some missing values).
Mean (Average Value): 104.8 → Typical measurement reading.
Minimum Value: -16.39 → Possible outlier (negative values in water quality?).
Maximum Value: 1,470,000 → Very high, needs checking.
    
Data Spread:
25% of values are below 1.25.
50% (Median) of values are below 6.97.
75% of values are below 18.0.
Standard Deviation (Std): 2,752.47 → Extremely high variation.
    
Key Insight:
✅ Possible data quality issues due to extreme outliers.
✅ Most readings are below 20, but max = 1.47M suggests anomalies.

In [47]:
# seeing duplicate rows and handling duplicates
duplicate_rows = df[df.duplicated(keep=False)]
print(f"Duplicate rows found: {len(duplicate_rows)}")
if not duplicate_rows.empty:
    print("Displaying duplicate rows:")
    print(duplicate_rows.sort_values(by=['Sample Id', 'Sample Date', 'Sample Time']))
    user_decision = input("Do you want to remove all duplicate rows? (yes/no): ")
    if user_decision.lower() == 'yes':
        df.drop_duplicates(inplace=True)
        print(f"Duplicates removed. New Shape: {df.shape}")
    else:
        print("No duplicates were removed.")

Duplicate rows found: 16
Displaying duplicate rows:
            Sample Id Sample Site Sample Date         Sample Time  \
1580221  E-1901638-01         BG9  2019-04-17 1900-01-01 11:44:00   
1580239  E-1901638-01         BG9  2019-04-17 1900-01-01 11:44:00   
1580367  E-1901638-02          E9  2019-04-17 1900-01-01 10:32:00   
1580392  E-1901638-02          E9  2019-04-17 1900-01-01 10:32:00   
1580251  E-1901638-03         E10  2019-04-17 1900-01-01 10:50:00   
1580269  E-1901638-03         E10  2019-04-17 1900-01-01 10:50:00   
1580281  E-1901638-04         E11  2019-04-17 1900-01-01 11:22:00   
1580291  E-1901638-04         E11  2019-04-17 1900-01-01 11:22:00   
1580342  E-1901638-05        MB-1  2019-04-17 1900-01-01 13:18:00   
1580371  E-1901638-05        MB-1  2019-04-17 1900-01-01 13:18:00   
1580174  E-1901638-06         N12  2019-04-17 1900-01-01 12:22:00   
1580324  E-1901638-06         N12  2019-04-17 1900-01-01 12:22:00   
1580199  E-1901638-07        N5-1  2019-04-17 1900-

Do you want to remove all duplicate rows? (yes/no):  yes


Duplicates removed. New Shape: (1664699, 9)


In [25]:
# number of unique values in each columns
for col in df.columns:
    print(f"{col}: {df[col].nunique()} unique values")

Sample Id: 117078 unique values
Sample Site: 423 unique values
Sample Date: 8351 unique values
Sample Time: 807 unique values
Analyte: 348 unique values
Status: 5 unique values
Final Result: 13857 unique values
Units: 28 unique values
Stream Group: 16 unique values


In [48]:
#checking missing values percentage:
missing_percentage = (df.isnull().sum() / len(df))*100
print(missing_percentage.sort_values(ascending=False))

Sample Time     82.888859
Status          82.088233
Final Result    14.681994
Units            8.200882
Stream Group     0.192948
Sample Id        0.000000
Analyte          0.000000
Sample Site      0.000000
Sample Date      0.000000
dtype: float64


 Interpretation of Missing Data Percentage
 
Sample Time → 82.89% missing
🔴 Critical issue: Only ~17% of samples have a recorded time.
🔹 Suggestion: Drop this column if time is not important, or analyze missing patterns.

Status → 82.09% missing
🔴 Severely incomplete: Only ~18% of records have a status.
🔹 Suggestion: Check if missing values correlate with a certain type of data.

Final Result → 14.68% missing
🟡 Moderate issue: ~15% of sample measurements are missing.
🔹 Suggestion: Consider imputation (e.g., median value) or drop missing rows if necessary.

Units → 8.20% missing
🟡 Small issue: Units are missing in ~8% of cases.
🔹 Suggestion: Check if certain analytes are more affected.

Stream Group → 0.19% missing
✅ Minor issue: Less than 1% missing → Not a major concern.

Sample Id, Analyte, Sample Site, Sample Date → 0% missing
✅ Perfect! These columns have complete data.



In [49]:
#dropping columns with excessive missing values
drop_columns = ['Sample Time', 'Status']
df.drop(columns=drop_columns, inplace=True)
print(f"Dropped columns: {drop_columns}")

Dropped columns: ['Sample Time', 'Status']


In [54]:
#filling missing values without explicit assignment of inplace=True as it might support some panda versions in the future
df = df.assign(
    Units = df['Units'].fillna(df['Units'].mode()[0]),
    Stream_Group = df['Stream Group'].fillna(df['Stream Group'].mode()[0]),    
    Final_Result = df['Final Result'].fillna(df['Final Result'].median())
)

In [55]:
#checking for negative or extreme values:
print("Negative values in 'Final Result':")
print(df[df['Final Result'] < 0])

Negative values in 'Final Result':
            Sample Id Sample Site Sample Date      Analyte  Final Result  \
217           E-96871   MIDDLEBRR  1987-01-05  Temperature      -1.11111   
371           E-95643     EASTBRR  1987-01-05  Temperature      -2.22222   
499           D-69955         RD1  1987-01-06  Temperature      -1.00000   
623           D-70457         RD4  1987-01-06  Temperature      -2.00000   
835           E-98396     WESTBRR  1987-01-06  Temperature      -2.22222   
...               ...         ...         ...          ...           ...   
1545405  K-1800859-04       STHHG  2018-03-12  Temperature      -0.10000   
1548055  K-1801184-02        SSHG  2018-04-09  Temperature      -0.20000   
1548059  K-1801184-03        SSMA  2018-04-09  Temperature      -0.20000   
1602967  G-2000341-11         NCG  2020-01-27     fDOM RFU      -0.10000   
1607029  G-2000999-09         NCG  2020-03-16     fDOM RFU      -0.10000   

        Units   Stream Group   Stream_Group  Final_R

Found 2,480 rows with negative values in Final Result, which likely indicates errors, missing data representations, or incorrect measurements. 

In [58]:
#removing unncessary columns
drop_unnecessary = ['Sample Id']
df.drop(columns=drop_unnecessary, inplace=True)
print(f"Dropped unnecessary columns: {drop_unnecessary}")

KeyError: "['Sample Id'] not found in axis"

In [59]:
# saving the cleaned file
cleaned_file_path = "../data/cleaned_water_quality.csv"
df.to_csv(cleaned_file_path, index=False)
print(f"Data cleaning completed. Cleaned dataset saved to {cleaned_file_path}")
print(f"Final Dataset shape: {df.shape}")

Data cleaning completed. Cleaned dataset saved to ../data/cleaned_water_quality.csv
Final Dataset shape: (1664699, 8)
